pkg/device_plugin/device_plugin.go

```diff
@@ -161,214 +155,79 @@ func startVgpuDevicePluginFunc(dp *GenericVGpuDevicePlugin) error {
+
+	for _, dev := range devices {
+		if dev.Driver == "vfio-pci" {
+			log.Printf("Found NVIDIA device: address=%s, driver=%s, iommu_group=%d, deviceID=%x", dev.Address, dev.Driver, dev.IommuGroup, dev.Device)
+			iommuGroupStr := strconv.Itoa(dev.IommuGroup)
+			_, exists := iommuMap[iommuGroupStr]
+			if !exists {
+				deviceMap[dev.Device] = append(deviceMap[dev.Device], iommuGroupStr)
 			}
+			iommuMap[iommuGroupStr] = append(iommuMap[iommuGroupStr], (*NvidiaPCIDevice)(dev))
 		}
-		return nil
-	})
+	}
 }
 
 //Discovers all Nvidia vGPUs configured on a node and creates corresponding maps
-func createVgpuIDMap() {
-	vGpuMap = make(map[string][]NvidiaGpuDevice)
+func createVgpuIDMap(nvmdev nvmdev.Interface) {
+	vGpuMap = make(map[string][]*NvidiaMDEVDevice)
 	gpuVgpuMap = make(map[string][]string)
-	//Walk directory to discover vGPU devices
-	filepath.Walk(vGpuBasePath, func(path string, info os.FileInfo, err error) error {
-		if err != nil {
-			log.Printf("Error accessing file path %q: %v\n", path, err)
-			return err
-		}
-		if info.IsDir() {
-			log.Println("Not a device, continuing")
-			return nil
-		}
-		//Read vGPU type name
-		vGpuID, err := readVgpuIDFromFile(vGpuBasePath, info.Name(), "mdev_type/name")
-		if err != nil {
-			log.Println("Could not get vGPU type identifier for device ", info.Name())
-			return nil
-		}
-		//Retrieve the gpu ID for this vGPU
-		gpuID, err := readGpuIDForVgpu(vGpuBasePath, info.Name())
-		if err != nil {
-			log.Println("Could not get vGPU type identifier for device ", info.Name())
-			return nil
-		}
-		log.Printf("Gpu id is %s", gpuID)
-		log.Printf("Vgpu id is %s", vGpuID)
-		gpuVgpuMap[gpuID] = append(gpuVgpuMap[gpuID], info.Name())
-		vGpuMap[vGpuID] = append(vGpuMap[vGpuID], NvidiaGpuDevice{info.Name()})
-		return nil
-	})
-}
 
-//Read a file to retrieve ID
-func readIDFromFileFunc(basePath string, deviceAddress string, property string) (string, error) {
-	data, err := ioutil.ReadFile(filepath.Join(basePath, deviceAddress, property))
+	devices, err := nvmdev.GetAllDevices()
 	if err != nil {
-		glog.Errorf("Could not read %s for device %s: %s", property, deviceAddress, err)
-		return "", err
+		// If the vGPU host driver is not installed, it is possible
+		// that /sys/bus/mdev/devices does not exist on the system.
+		// Return early but do not treat this as a critical error.
+		log.Printf("Error getting NVIDIA vGPU devices: %v", err)
+		return
 	}
-	id := strings.Trim(string(data[2:]), "\n")
-	return id, nil
-}
 
-//Read a file link
-func readLinkFunc(basePath string, deviceAddress string, link string) (string, error) {
-	path, err := os.Readlink(filepath.Join(basePath, deviceAddress, link))
-	if err != nil {
-		glog.Errorf("Could not read link %s for device %s: %s", link, deviceAddress, err)
-		return "", err
-	}
-	_, file := filepath.Split(path)
-	return file, nil
-}
+	for _, dev := range devices {
+		gpuAddress := dev.Parent.Address
+		vgpuType := strings.ReplaceAll(dev.MDEVType, " ", "_")
```

Comment:
Note, before this change, the device plugin would use the full contents of the `mdev_type/name` file to construct the corresponding resource name. E.g. `mdev_type/name` would contain  the string `NVIDIA A10-12Q` and the plugin would advertise `nvidia.com/NVIDIA_A10-12Q` resources. The nvmdev package strips the leading `NVIDIA | GRID` in this file, so the resource name would now be `nvidia.com/A10-12Q`. 

I personally would prefer the latter naming strategy, but this is a breaking change from the user's perspective. I have opened https://gitlab.com/nvidia/cloud-native/go-nvlib/-/merge_requests/45 to add a `device.MDEVName` field so we can use the full contents of `mdev_type/name` when constructing the resource name (and thereby retain the current behavior of the plugin).